src/libs/elektra/changetracking.c

```diff
@@ -0,0 +1,601 @@
+#include <kdbchangetracking.h>
+#include <kdbprivate.h>
+
+/**
+ * Returns the changetracking context of the given KDB instance
+ *
+ * @param kdb the KDB instance
+ * @return ChangeTrackingContext or @p NULL if @p kdb is NULL
+ */
+const ChangeTrackingContext * elektraChangeTrackingGetContextFromKdb (KDB * kdb)
+{
+	if (kdb == NULL)
+	{
+		return NULL;
+	}
+
+	return &kdb->changeTrackingContext;
+}
+
+/**
+ * Returns the changetracking context for the KDB instance associated with the specified plugin
+ *
+ * @param plugin the plugin
+ * @return ChangeTrackingContext or @p NULL if @p plugin is NULL or does not have a @p KDB instance associated with it
+ */
+const ChangeTrackingContext * elektraChangeTrackingGetContextFromPlugin (Plugin * plugin)
+{
+	if (plugin->global == NULL)
+	{
+		return NULL;
+	}
+
+	Key * kdbKey = ksLookupByName (plugin->global, "system:/elektra/kdb", 0);
+	if (kdbKey == NULL)
+	{
+		return NULL;
+	}
+
+	const void * kdbPtr = keyValue (kdbKey);
+	KDB * kdb = kdbPtr == NULL ? NULL : *(KDB **) keyValue (kdbKey);
+
+	return elektraChangeTrackingGetContextFromKdb (kdb);
+}
+
+/**
+ * @internal
+ *
+ * @brief For testing purposes only: Create a ChangeTrackingContext from a keyset
+ *
+ * @param oldKeys the old keys
+ * @return a new changetracking context
+ */
+ChangeTrackingContext * elektraChangeTrackingCreateContextForTesting (KeySet * oldKeys)
+{
+	ChangeTrackingContext * context = elektraCalloc (sizeof (ChangeTrackingContext));
+
+	ksIncRef (oldKeys);
+	context->oldKeys = oldKeys;
+
+	return context;
+}
```

Comment:
Do we need this here? Couldn't this be in the tests somewhere? If it's needed in multiple places, you could create a header `src/include/kdbchangetracking_tests.h` with:

```c
static inline ChangeTrackingContext * elektraChangeTrackingCreateContextForTesting (KeySet * oldKeys)
{
	ChangeTrackingContext * context = elektraCalloc (sizeof (ChangeTrackingContext));

	ksIncRef (oldKeys);
	context->oldKeys = oldKeys;

	return context;
}
```

Reply:
But then the problem would be that the tests themselves would need to have access to the internal structure of the `ChangeTrackingContext` struct.

Reply:
That's not really an issue in the tests. You can just include the private headers.